### PR TITLE
Rebase and fixup #80493: Remove MIR assignments to ZST types

### PR DESCRIPTION
**Pull request by erikdesjardins**
closes #80493

cc @simonvandel

r? @oli-obk

### CODE REVIEW

**Comment by erikdesjardins** [comment]
Hmm, does pushing mess up the try build? I thought it would be fine as long as I didn't force-push

File: compiler/rustc_mir/src/transform/simplify.rs [hunk 1]
```diff
@@ -451,7 +459,23 @@ impl Visitor<'_> for UsedLocals {
         }
     }
 
-    fn visit_local(&mut self, local: &Local, _ctx: PlaceContext, _location: Location) {
+    fn visit_local(&mut self, local: &Local, ctx: PlaceContext, _location: Location) {
+        debug!("local: {:?} is_static: {:?}, ctx: {:?}", local, self.is_static, ctx);
+        // Do not count a local as used in `_local = <rhs>` if RHS is a ZST.
+        let store = matches!(ctx, PlaceContext::MutatingUse(MutatingUseContext::Store));
```
Comment (tmiasko) [inline compiler/rustc_mir/src/transform/simplify.rs:465]: The essence of the current algorithm is that it skips visitation of a `_local` in `_local = <rhs>` statements. This condition should always be false, except for an inline assembly where visitation process is not customized.

Comment (erikdesjardins) [reply compiler/rustc_mir/src/transform/simplify.rs]: Yep, tests still pass locally with this removed

**Comment by oli-obk** [comment]
@bors r+

Thanks for reviving that PR!

**Comment by tmiasko** [comment]
That landed fast. I will just note that it might be interesting to explore removal of individual ZST assignments still, rather than locals, since this miss the opportunities that exists around code that is using `?` with ZST types (`?` introduces var debug info and const propagation doesn't propagate into call terminators by default, so locals involved remain because they are used).

**Comment by erikdesjardins** [comment]
That's what #80493 originally did before https://github.com/rust-lang/rust/pull/80493#issuecomment-752668223: see https://github.com/rust-lang/rust/compare/22f94499cc9c397420c4cf8ba0bbcabf4018bae7

I'll open a PR.

**Pull request merged**